rustls/src/msgs/deframer.rs

```diff
@@ -462,15 +491,28 @@ impl<'a> DeframerSliceBuffer<'a> {
     pub fn is_empty(&self) -> bool {
         self.len() == 0
     }
+
+    fn take(&mut self, raw: RawSlice) -> &'a mut [u8] {
+        let start = (raw.ptr as usize)
+            .checked_sub(self.buf.as_ptr() as usize)
+            .unwrap();
```

Comment:
So we're doing pointer arithmetic here -- how do we guarantee that `raw.ptr` and `self.buf` are from the same allocation? Even though there is no `unsafe` code here, this seems to be potentially brittle code which should come with more documentation.

Reply:
I'm struggling to understand the `RawSlice` abstraction and the commit that introduced it is missing any motivating context or guidance in the commit message (I guess that's part of what you're flagging in this review...). I gather we want to use pointers here instead of a normal slice into `self.buf` for reasons related to the lifetime of the slice and the `DeframerSliceBuffer` buffer, but I can't exactly put my finger on it.

>  how do we guarantee that raw.ptr and self.buf are from the same allocation?

I haven't been able to think of a way to tie the `RawSlice` back to the buffer it came from that won't trip into the lifetime issues that I think this design was built around avoiding. Do you have any suggestions?

>  this seems to be potentially brittle code which should come with more documentation.

Would additional documentation about the invariants be enough to satisfy your concerns? That's certainly doable :+1:

Reply:
> how do we guarantee that raw.ptr and self.buf are from the same allocation?

We could drop the `RawSlice::from<&[u8]>` constructor and implement a method on `DeframerSliceBuffer` that takes a slice argument and asserts the base ptr and len of the slice falls within the `DeframerSliceBuffer`'s `buf` before giving back the `RawSlice`.

Writing that out, I'm not sure that it's a very useful change w.r.t safety or brittleness since it just changes where we panic: constructing the `RawSlice` or doing the `split_at_mut`/`checked_sub` in `take`/`raw_slice_to_filled_range` with the result of the pointer arithmetic.

@ctz Do you have any thoughts on this feedback or suggestions for refactoring?

Reply:
I'd say we're doing a similar thing in a different way elsewhere inside msgs/deframer.rs: but with `Range` rather than pointer arithmetic.

That is, at least, more obvious -- though I think it still suffers from similar fragility problems if anyone changes the buffer while the `Range` is valid. That instance is especially egregious because the buffer is under control of the application while the `Range` is valid. I guess this is one of the negative results of pursuing a 100% (including handshaking) no-allocation solution here rather than just addressing the main data path.

Reply:
I have an extremely broken and half-baked branch with some refactoring in this area, but haven't reached an ending on it (either achieving something good, or giving up!)

Some of the ideas I've so far failed to achieve:

- The deframer only ever reads from its buffer, and the `Deframed` object it returns doesn't keep a borrow of the message data; it just tells you where it is. That gives the deframer potential to become an iterator over `Result<Deframed, Error>` items in the future, maybe.
- Someone with a `Deframed` message can decrypt it in-place by providing the `record_layer` and the mutable slice containing it. That process should also not keep a mutable borrow of the buffer.
- The handshake part of the deframer is split out into a separate thing. At the moment the deframer is doing both slicing up TLS messages, and also gluing together TLS handshake protocol fragments. I would hope this smooths out the complexity hotspot in the deframer.
- The QUIC code can directly use the handshake-part-of-the-deframer, which should relieve us of a lot of the `if quic { }` decisions in the deframer.

Reply:
That sounds great. Maybe @cpu could take the half-baked branch and run with it?

Reply:
Pushed https://github.com/rustls/rustls/tree/jbp-refactor-deframer

Reply:
> That sounds great. Maybe @cpu could take the half-baked branch and run with it?

Sure, I can give that a go :+1:, but is it a blocker for landing this work? I'd rather not have to maintain two unfinished branches at the same time if possible. If it's OK with you folks my preference would be to land this branch and then rebase the refactor on top to continue.

> Pushed https://github.com/rustls/rustls/tree/jbp-refactor-deframer

@ctz Could you maybe summarize where you left off to help me get started?

Reply:
> That seems reasonable.

Thanks :+1: I'll add some additional documentation around the `RawSlice` invariants and then merge this.

Reply:
Updated with additional doc comments.